bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,479 @@
+    /// wait for a tick.
+    ///
+    /// This call can never trigger a state transition, so we don't bother
+    /// taking `self` by value.
+    fn on_disconnect(
+        &mut self,
+        common: &mut FsmCommonData,
+        peer: Baseboard,
+    ) -> Output;
+}
+
+/// The state machine for a [`$crate::Peer`]
+///
+/// This FSM assumes a network layer above it that can map peer IDs
+/// ( [`Baseboard`]s) to TCP sockets. When an attempt is made to send a
+/// message to a given peer the network layer will send it over an established
+/// connection if one exists. If there is no connection already to that peer,
+/// but the prefix is present and known, the network layer will attempt to
+/// establish a connection and then transmit the message. If the peer is not
+/// known then it must have just had its network prefix removed. In this case
+/// the message will be dropped and the FSM will be told to remove the peer.
+#[derive(Debug)]
+pub struct Fsm {
+    common: FsmCommonData,
+
+    // Use an option to allow taking and mutating `State` independently of `Fsm`
+    // An invariant of the Fsm is that `state` is always `Some` to API callers.
+    state: Option<State>,
+}
+
+impl Fsm {
+    /// Create a new FSM. This is useful when starting from a persistent state
+    /// on disk.
+    pub fn new(id: Baseboard, config: Config, state: State) -> Fsm {
+        Fsm { common: FsmCommonData::new(id, config), state: Some(state) }
+    }
+
+    /// Create a new FSM in `State::Uninitialized`
+    pub fn new_uninitialized(id: Baseboard, config: Config) -> Fsm {
+        let state = Some(State::Uninitialized(UninitializedState {}));
+        Fsm { common: FsmCommonData::new(id, config), state }
+    }
+
+    pub fn state_name(&self) -> &'static str {
+        self.state.as_ref().unwrap().name()
+    }
+
+    pub fn state(&self) -> &State {
+        self.state.as_ref().unwrap()
+    }
+
+    pub fn common_data(&self) -> &FsmCommonData {
+        &self.common
+    }
+
+    /// This call is triggered locally as a result of RSS running
+    /// It may only be called once, which is enforced by checking to see if
+    /// we already are in `State::Uninitialized`.
+    pub fn init_rack(
+        &mut self,
+        rack_uuid: Uuid,
+        initial_membership: BTreeSet<Baseboard>,
+    ) -> Output {
+        let State::Uninitialized(_) = self.state.as_ref().unwrap() else {
+            return ApiError::RackAlreadyInitialized.into();
+        };
+        let total_members = initial_membership.len();
+        match create_pkgs(rack_uuid, initial_membership.clone()) {
+            Ok(pkgs) => {
+                // We learn our pkg by iterating
+                let mut our_pkg = None;
+                let request_id = Uuid::new_v4();
+                let envelopes: Vec<_> = pkgs
+                    .expose_secret()
+                    .into_iter()
+                    .zip(initial_membership)
+                    .filter_map(|(pkg, peer)| {
```

Comment:
Does preserving order of `initial_membership` actually matter here?  The code would be a bit cleaner if we could claim the first package for ourself, e.g.
```rust
                let mut iter = pkgs.expose_secret().into_iter();
                let our_pkg = iter.next().unwrap().clone();

                let envelopes: Vec<_> = initial_membership
                    .into_iter()
                    .filter(|peer| *peer != self.common.id)
                    .zip(iter)
                    .map(|(peer, pkg)| Envelope {
                        to: peer,
                        msg: Request {
                            id: request_id,
                            type_: RequestType::Init(pkg.clone()),
                        }
                        .into(),
                    })
                    .collect();
```

(this passes the proptest)

Reply:
Cleaned up a bit with the restructuring in [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98)